Feature analysis tools

lomamech opened the pull request:
This PR Implementing a set of support tools to assess the institution's ability to recover its revenues and to cover its costs and debts, the report resulting from this module will be an implementation computer report attached

lomamech committed:
improvement(Analysis Tools)
- Improved reporting, added the ability to include Posting Journal Data where
  possible

lomamech committed:
feature(Analysis Tools)
- Prevent the consultation of the Analysis Report when the necessary
  configurations are missing
- Adding of the option to include and exclude data from the Posting Journal

lomamech committed:
Resolve Conflict

lomamech committed:
Clear and Sanitaze Code

lomamech committed:
Fix error in e2e file for Configuration Analysis tools

lomamech committed:
test(Account Reference)
- Reset default account reference number

lomamech committed:
test(Account Reference)
- Reset default account reference number in search

lomamech committed:
test(Account Reference Type)
- Update number of account reference type in E2E test

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+// default report parameters
+const DEFAULT_PARAMS = {
+  csvKey : 'brea_report',
+  filename : 'TREE.BREAK_EVEN_REPORT',
```
Is this the `breakeven` analysis report?

jniles commented on client/src/modules/reports/generate/configurable_analysis_report/configurable_analysis_report.html:
```diff
@@ -0,0 +1,77 @@
+            </bh-fiscal-period-select>
+
+            <hr>
+            <bh-multiple-cashbox-select
```
I'm not sure if this is a problem with the form or the component, but the input is not highlighted in red when it is required.

jniles reviewed (commented):
I tried running this report, and I got the following error message:

![image](https://user-images.githubusercontent.com/896472/67752589-86999900-fa33-11e9-9655-1b576e550385.png)


Do you know what this is about?  All I did was run the integration tests and try to render the report.

mbayopanda commented on server/models/migrations/next/migrate.sql:
```diff
@@ -5,3 +5,60 @@ ALTER TABLE `stock_consumption` MODIFY `period_id` MEDIUMINT(8) UNSIGNED NOT NUL
+INSERT INTO `analysis_tool_type` (`id`, `label`, `balance_type`, `rank`, `is_balance_sheet`) VALUES (1, 'FORM.LABELS.ANALYSIS_TOOLS.COSTS', 'FORM.LABELS.DEBIT_BALANCE', 1, 0);
+INSERT INTO `analysis_tool_type` (`id`, `label`, `balance_type`, `rank`, `is_balance_sheet`) VALUES (2, 'FORM.LABELS.ANALYSIS_TOOLS.RECEIVABLES', 'FORM.LABELS.DEBIT_BALANCE', 4, 1);
+INSERT INTO `analysis_tool_type` (`id`, `label`, `balance_type`, `rank`, `is_balance_sheet`) VALUES (3, 'FORM.LABELS.ANALYSIS_TOOLS.PROFITS', 'FORM.LABELS.CREDIT_BALANCE', 2, 0);
+INSERT INTO `analysis_tool_type` (`id`, `label`, `balance_type`, `rank`, `is_balance_sheet`) VALUES (4, 'FORM.LABELS.ANALYSIS_TOOLS.DEBTS', 'FORM.LABELS.CREDIT_BALANCE', 3, 1);
```
You have to remove these duplicated lines (57 to 60)

mbayopanda commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+      `;
+
+      const sqlReferences = `
+        SELECT at.id, at.is_creditor, at.restrict_title_account, at.account_reference_id,
```
The column `restrict_title_account` doesn't exist in the configuration_analysis_tools table.

mbayopanda commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+
+      const sqlConfig = `
+        SELECT id, is_creditor, restrict_title_account, account_reference_id, label,
+        analysis_tool_type_id
```
The column `restrict_title_account` doesn't exist in the configuration_analysis_tools table.

mbayopanda commented on server/controllers/finance/reports/configurable_analysis_report/report.handlebars:
```diff
@@ -0,0 +1,154 @@
+                      <td width="59%" rowspan="2" colspan="4" class="text-uppercase text-bold"> {{translate label}} </td>
+                      <td class="text-center" width="13%"> {{translate 'FORM.LABELS.DEBIT'}} </td>
+                      <td class="text-center" width="13%"> {{translate 'FORM.LABELS.CREDIT'}} </td>
+                      <td class="text-center" width="13%"> {{translate displayLabel}} </td>  
```
Is it possible to remove all these repetitive `debit`, `credit` and `balance`, and just keep which are on the top of all sections

lomamech replied:
I tried to remove it but it may confuse the headers of each subcategory, and it also avoids that we can be forced to search the first page to know the title of the headers

mbayopanda commented on server/models/bhima.sql:
```diff
@@ -362,4 +366,12 @@ INSERT INTO `cron` (`label`, `value`) VALUES
-  ('CRON.YEARLY', '0 1 31 12 *');
\ No newline at end of file
+  ('CRON.YEARLY', '0 1 31 12 *'),
+  ('CRON.EACH_MINUTE', '* * * * *');
```
this line was removed, could you remove it

mbayopanda commented on server/models/migrations/next/migrate.sql:
```diff
@@ -5,3 +5,60 @@ ALTER TABLE `stock_consumption` MODIFY `period_id` MEDIUMINT(8) UNSIGNED NOT NUL
+  `label` VARCHAR(100) NOT NULL,
+  `is_creditor` tinyint(1) DEFAULT 0,
+  `account_reference_id` MEDIUMINT(8) UNSIGNED NOT NULL,
+  `analysis_tool_type_id` MEDIUMINT(8) UNSIGNED NOT NULL,
```
the column `restrict_title_account` is missing

lomamech replied:
This property was no longer useful, it was deleted, in the next PR I will cleaned all its occurrences

mbayopanda reviewed (commented):
i made some comments about some repetitive `debit`, `credit` and `balance` in the report template, i think in this way the report will be more clean

lomamech committed:
Fix: remove all occurence of restrict_title_account

lomamech committed:
improvement(Component)
- Set Input highlighted in red when it's required for bhFiscalPeriodSelect
  and bhMultipleCashboxSelect

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+ * @description
+ *
+ */
+function report(req, res, next) {
```
This controller is super hard to understand because there are no comments.  Could you comment your code to explain to us what is happening?

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+  const cashboxesIds = _.values(req.query.cashboxesIds);
+
+  data.options = {
+    display_account_details : parseInt(params.hide_account_details, 10),
```
Why is this `display_account_details` and not `hide_account_details`?  It seems like you inverted the meaning of the boolean.

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+  };
+
+  params.start_date = new Date(params.start_date);
+  data.currencyId = params.currency_id;
```
In the report page, there isn't an option for selecting a currency.  Where does this currency come from?

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+    fiscal_year_id : params.fiscalYearId,
+  };
+
+  params.includeUnpostedValues = parseInt(params.includeUnpostedValues, 10);
```
Where did this come from?  There isn't an option for it on the report page.

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+    .then(rows => {
+      const getOpening = [];
+      data.cashboxes = rows;
+      data.cashboxes.forEach(cash => {
```
Anytime there is code that looks like this:
```js
const a = [];
b.forEach(v => {
  a.push(v);
});
return a;
```
you are actually doing a [map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).  Please just use a map instead.  Like this:
```js
.then(rows => {
  const promises = rows.map(row => AccountExtras.getOpeningBalanceForDate(row.account_id, data.period.end_date));
  return q.all(promises);
});
```

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+        JOIN period AS p ON p.id = pt.period_id
+        JOIN account AS a ON a.id = pt.account_id
+        WHERE (DATE(p.end_date) < DATE(?) OR p.number = 0) AND pt.fiscal_year_id = ?
+        AND a.type_id <= 3 AND a.id IN (?)
```
Why is this `a.type_id <= 3`?  What is type 3?  Why can't we have `id`s less than 3?

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+      ];
+
+      const sqlOpeningBalanceSheet = `
+        SELECT (a.id) AS account_id, a.number, p.translate_key, a.label, SUM(pt.debit) AS debit,
```
You don't need the `(a.id)` parentheses.  I think you can just do `SELECT a.id, ...`.

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+                  AND DATE(i.date) >= DATE(?)
+              ) AS rev
+            )
+            UNION
```
This should be `UNION ALL`.  The `UNION` clause will try to remove duplicates, which affects performance.

jniles commented on server/controllers/finance/reports/configurable_analysis_report/index.js:
```diff
@@ -0,0 +1,382 @@
+        });
+
+        balance.forEach(item => {
+          const accountNumber = item.number.toString();
```
I though the point of using account references was so that we didn't have to parse the account number.  Does this mean we now have to tell our users they can't pick their own account numbers?

jniles reviewed (commented):
A few comments here.  We spoke in person about switching the sign for debit or credit balances.

lomamech committed:
Rebase with branch Master

lomamech committed:
Fix and resolve conflict

lomamech committed:
Rebase with branch master

lomamech committed:
Resolve conflict with branch master

lomamech committed:
Complete resolve conflict

mbayopanda commented on client/src/modules/configuration_analysis_tools/configuration_analysis_tools.js:
```diff
@@ -0,0 +1,116 @@
+
+        });
+
+        vm.gridOptions.data = configurationAnalysisTools;
```
I think you could also do : 
```js
vm.gridOptions.data = configurationAnalysisTools.map(item => {
  if (item.typeLabel) {
    item.typeLabel = $translate.instant(item.typeLabel);
  }
  return item;
});
```

mbayopanda commented on client/src/modules/reports/generate/configurable_analysis_report/configurable_analysis_report.config.js:
```diff
@@ -0,0 +1,98 @@
+  const reportUrl = 'reports/finance/configurable_analysis_report';
+
+  ConfigurationAnalysisTools.read()
+    .then((configurationAnalysisTools) => {
```
Just FYI, with arrow-function if you have just one parameter you can write the arrow-function without parenthesis for parameter.
```js
.then(configurationAnalysisTools => {
```

mbayopanda commented on client/src/modules/templates/bhMultipleCashBoxSelect.tmpl.html:
```diff
@@ -1,7 +1,7 @@
   <div
     class="form-group"
-    ng-class="{ 'has-error' : $ctrl.validationTrigger && CashboxForm.cashbox.$invalid }">
+    ng-class="{ 'has-error' : CashboxForm.$submitted && CashboxForm.cashbox.$invalid }">
```
👍

mbayopanda commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "dev": "cross-env NODE_ENV=development gulp build && cd bin && cross-env NODE_ENV=development node server/app.js",
     "test": "./sh/test.sh",
-    "test:integration": "SUITE=\"integration\" ./sh/test.sh",
+    "test:integration": "cross-env SUITE=\"integration\" ./sh/test.sh",
```
👍

mbayopanda commented on server/controllers/finance/configurationAnalysisTools.js:
```diff
@@ -0,0 +1,135 @@
+/**
+* Configuration Analysis Tools Controller
+*
+* This controller exposes an API to the client for reading and writing Break Even Reference
```
This controller is for `Break Even Reference` ? if not, could you fix this comment

mbayopanda commented on server/controllers/finance/configurationAnalysisTools.js:
```diff
@@ -0,0 +1,135 @@
+  const sql = `
+    SELECT id, label, account_reference_id, analysis_tool_type_id
+    FROM configuration_analysis_tools
+    WHERE configuration_analysis_tools.id = ?`;
```
`configuration_analysis_tools.id = ?` is longer than just `id = ?`

mbayopanda commented on server/controllers/finance/configurationAnalysisTools.js:
```diff
@@ -0,0 +1,135 @@
+  db.exec(sql)
+    .then((rows) => {
+      res.status(200).json(rows);
+    })
```
Just FYI, do you know that you can write : 
```js
.then(rows => res.status(200).json(rows))
.catch(next)
.done();
```

mbayopanda reviewed (approved):
I made some non-blocking comments, i am just wondering if you have applied @jniles comments

LGTM 👍

lomamech committed:
Fix conflict

lomamech committed:
Rebase with branch Master

lomamech committed:
Fix conflict

lomamech committed:
Fix and Resolve conflict

lomamech committed:
Resolve conflict

lomamech committed:
Restaure file package.json

lomamech committed:
test: refactor value for stock Movemenet E2E test

lomamech committed:
Test: reset value for E2E test

lomamech committed:
Test: refactor number of row
